### PR TITLE
Move to Github Actions

### PR DESCRIPTION
**Pull request by alain-andre**
This moves CI from Travis to GitHub actions. I forced the tests to start only on dev branch like it was asked before.

I would prefer to launch tests. For example this PR does not show the tests results you can see on this [branch](https://github.com/alain-andre/optimizer-api/tree/dev)

### CODE REVIEW

**Comment by fonsecadeline** [comment]
When you say 

> I forced the tests to start only on dev branch 

does it mean that if I push on a branch called _hot_fix_ then no integration tests will be run ? Is that what we want ?

File: README.md [hunk 1]
```diff
@@ -1,4 +1,5 @@
-# Mapotempo Optimizer API [![Build Status](https://travis-ci.com/Mapotempo/optimizer-api.svg?branch=master)](https://travis-ci.com/Mapotempo/optimizer-api)
+# Mapotempo Optimizer API ![dev Build Status](https://github.com/Mapotempo/optimizer-api/actions/workflows/main.yml/badge.svg?branch=dev)
```
Comment (fab-girard) [inline README.md:1]: why do you replace master branch by dev branch?

File: .github/actions/tests.sh [hunk 2]
```diff
@@ -0,0 +1,4 @@
+#!/usr/bin/env bash
+
+CONTAINER=${PROJECT}_api.1.$(docker service ps -f "name=${PROJECT}_api.1" ${PROJECT}_api -q --no-trunc | head -n1)
+docker exec -i ${CONTAINER} bundle exec rake test ${TEST_ENV} ${OPTIONS}
```
Comment (fab-girard) [inline .github/actions/tests.sh:4]: what is this env var `TEST_ENV`?

**Comment by senhalil** [comment]
If we have to choose, I think we should run the tests on PR's so that we don't merge a broken PR. 

Ideally we would run the tests on every branch and not just `dev` branch, this was what we did on Travis-CI. Since we do our developments on different branches (so that we can keep the PR's separate), it is important to run the tests on other branches.

File: README.md (hunk 1 above)
Comment (alain-andre) [reply README.md:1]: I thought it was interesting to specify the dev, but I will move it back to master branch.

**Comment by alain-andre** [comment]
> When you say
> 
> > I forced the tests to start only on dev branch
> 
> does it mean that if I push on a branch called _hot_fix_ then no integration tests will be run ? Is that what we want ?

As seen together, it was a misunderstanding, I change this to be run on any branch.

File: .github/actions/tests.sh (hunk 2 above)
Comment (alain-andre) [reply .github/actions/tests.sh:4]: This comes from the old script, I remove it.

**Review by fonsecadeline — APPROVED** [review]
I do not have a great understanding of this but it seems good to me (mainly main.yml)

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -0,0 +1,145 @@
+name: Docker Image CI
+
+on:
+  push:
+
+env:
+  NB_SERVICES: ${{ secrets.NB_SERVICES }}
+  OPTIMIZER_ORTOOLS_VERSION: ${{ secrets.OPTIMIZER_ORTOOLS_VERSION }}
```
Comment (braktar) [inline .github/workflows/main.yml:8]: Do we have a way to put a default version ?

Comment (alain-andre) [reply .github/workflows/main.yml:8]: It uses the default version specified in Dockerfile. Those variables allow you to override the values.

Comment (senhalil) [reply .github/workflows/main.yml:8]: These secrets, are they like hidden environment variables on github side or they are defined inside our repo?

File: .github/workflows/main.yml [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+      - uses: actions/checkout@v1
+      - name: Cache
+        id: cache
+        uses: actions/cache@v2
+        with:
+          path: vendor/bundle
+          key: ${{ runner.os }}
+      - name: Build image
+        run: ./.github/actions/build.sh
+        shell: bash
```
Comment (senhalil) [inline .github/workflows/main.yml:28]: Can't we create the image just once and launch the tests after? I mean that looks more like proper workflow unless it is complicated.

Comment (alain-andre) [reply .github/workflows/main.yml:28]: We cannot create the image and pass it to another job. At every job, or we build it, or we download it.

File: .github/workflows/main.yml (hunk 3 above)
Comment (alain-andre) [reply .github/workflows/main.yml:8]: Secrets are defined inside your repo in the GitHub settings : `settings/secrets/actions`

File: .github/workflows/main.yml (hunk 4 above)
Comment (senhalil) [reply .github/workflows/main.yml:28]: So there is no way to set a common "environment" for all jobs :/ weird. Thanks Alain!

File: .github/actions/build.sh [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+  nbps=$(docker service ls | grep 1/1 | awk '{print $4}' | wc -l)
+  if [ ${nbps} -eq ${nb_services} ];
+  then
+    echo "All services up, starting tests"
+    break;
+  fi
+
+  echo "Waiting for all services (${nbps}/${nb_services}) to start ($cpt secondes)."
+  sleep 1
+done
```
Comment (senhalil) [inline .github/actions/build.sh:42]: Do we need this "try/fail" logic for Github as well?

**Review by senhalil — APPROVED** [review]

File: .github/actions/build.sh (hunk 5 above)
Comment (alain-andre) [reply .github/actions/build.sh:42]: Yes we need it to be sure all services are up before starting tests.

**Review by braktar — APPROVED** [review]

**Pull request merged**